segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/LogMetadata.java

```diff
@@ -140,6 +141,63 @@ LogMetadata truncate(LedgerAddress upToAddress) {
                 .withUpdateVersion(this.updateVersion.get());
     }
 
+    /**
+     * Removes LedgerMetadatas for those Ledgers that are known to be empty.
+     *
+     * @param skipCountFromEnd The number of Ledgers to spare, counting from the end of the LedgerMetadata list.
+     * @return A new instance of LogMetadata with the updated ledger list.
+     */
+    LogMetadata removeEmptyLedgers(int skipCountFromEnd) {
+        val newLedgers = new ArrayList<LedgerMetadata>();
+        int cutoffIndex = this.ledgers.size() - skipCountFromEnd;
+        for (int i = 0; i < cutoffIndex; i++) {
+            LedgerMetadata lm = this.ledgers.get(i);
+            if (lm.getStatus() != LedgerMetadata.Status.Empty) {
+                // Not Empty or Unknown: keep it!
+                newLedgers.add(lm);
+            }
+        }
+
+        // Add the ones from the end, as instructed.
+        for (int i = cutoffIndex; i < this.ledgers.size(); i++) {
+            newLedgers.add(this.ledgers.get(i));
+        }
+
+        return new LogMetadata(this.epoch, Collections.unmodifiableList(newLedgers), this.truncationAddress)
+                .withUpdateVersion(this.updateVersion.get());
+    }
+
+    /**
+     * Updates the LastAddConfirmed on individual LedgerMetadatas based on the provided argument.
+     *
+     * @param lastAddConfirmed A Map of LedgerId to LastAddConfirmed based on which we can update the status.
+     * @return This (unmodified) instance if lastAddConfirmed.isEmpty() or a new instance of the LogMetadata class with
+     * the updated LedgerMetadatas.
+     */
+    LogMetadata updateLedgerStatus(Map<Long, Long> lastAddConfirmed) {
+        if (lastAddConfirmed.isEmpty()) {
+            // Nothing to change.
+            return this;
+        }
+
+        val newLedgers = this.ledgers.stream()
+                .map(lm -> {
+                    long lac = lastAddConfirmed.getOrDefault(lm.getLedgerId(), Long.MIN_VALUE);
+                    if (lm.getStatus() == LedgerMetadata.Status.Unknown && lac != Long.MIN_VALUE) {
+                        LedgerMetadata.Status e = lac == Ledgers.NO_ENTRY_ID
```

Comment:
If I remember correctly, there is a corner case here. If we have written exactly one entry, then the last add confirmed will be no entry because the bookies need a second entry to confirm the first.

Reply:
My unit tests write exactly one entry, so this must work.

Reply:
The bk client updates its local copy of `lastAddConfirmed`, so if we are getting the value from the same handle that is being used to write, then it is fine.

The second case is when we open a closed ledger. When we open a closed ledger, the ledger handle will be initialized with the entry id in the ledger metadata, so this case is also fine.

The third case is a handle for an open ledger. In the case, the `lastAddConfirmed` value of the ledger handle is `INVALID_ENTRY_ID`.

Could you confirm that all ledger handles in `this.ledgers` are such that they are either closed or are the ones being used to add entries to the ledger?

Reply:
The unit test (check out `BooKeeperLogTests.java`) does this:
- Loop around 100 times; at every loop
    - Initialize the log (new ledger added)
    - Every 10 such initializations, write exactly one entry
    - Close the log
- Every time it opens the log, it verifies the metadata is as expected (empty ledgers only are removed) as well as the BK ledgers themselves are removed.

So yes, all ledgers in here are all previously closed (normally or fenced out) or the one that is currently open for writes (that is the last one). However, this method is called prior to creating the new ledger, so all the ledgers in here are previously closed and their LAC is fetched upon opening.

Reply:
Ok.